File: 2023/day05/ruby/solution.rb
# Day 5: If You Give A Seed A Fertilizer

def parse_input(text)
  sections = text.strip.split("\n\n")

  # Parse seeds
  seeds = sections[0].split(': ')[1].split.map(&:to_i)

  # Parse maps
  maps = sections[1..].map do |section|
    lines = section.strip.split("\n")
    lines[1..].map do |line|
      dst_start, src_start, length = line.split.map(&:to_i)
      [dst_start, src_start, length]
    end
  end

  [seeds, maps]
end

def apply_map(value, ranges)
  ranges.each do |dst_start, src_start, length|
    if value >= src_start && value < src_start + length
      return dst_start + (value - src_start)
    end
  end
  value
end

def seed_to_location(seed, maps)
  value = seed
  maps.each do |map_ranges|
    value = apply_map(value, map_ranges)
  end
  value
end

def part1(seeds, maps)
  seeds.map { |seed| seed_to_location(seed, maps) }.min
end

def apply_map_to_ranges(input_ranges, map_ranges)
  result = []

  input_ranges.each do |start, range_end|
    remaining = [[start, range_end]]

    map_ranges.each do |dst_start, src_start, length|
      src_end = src_start + length
      new_remaining = []

      remaining.each do |r_start, r_end|
        # Part before the map range (unmapped)
        if r_start < src_start
          new_remaining << [r_start, [r_end, src_start].min]
        end

        # Part within the map range (mapped)
        overlap_start = [r_start, src_start].max
        overlap_end = [r_end, src_end].min
        if overlap_start < overlap_end
          offset = dst_start - src_start
          result << [overlap_start + offset, overlap_end + offset]
        end

        # Part after the map range (unmapped)
        if r_end > src_end
          new_remaining << [[r_start, src_end].max, r_end]
        end
      end

      remaining = new_remaining
    end

    # Any remaining parts are unmapped (identity)
    result.concat(remaining)
  end

  result
end

def part2(seeds, maps)
  # Convert seeds to ranges: pairs of [start, start + length]
  ranges = seeds.each_slice(2).map do |start, length|
    [start, start + length]
  end

  # Apply each map to the ranges
  maps.each do |map_ranges|
    ranges = apply_map_to_ranges(ranges, map_ranges)
  end

  # Find minimum start of any range
  ranges.map(&:first).min
end

def main
  input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
  text = File.read(input_path)

  seeds, maps = parse_input(text)

  puts "Part 1: #{part1(seeds, maps)}"
  puts "Part 2: #{part2(seeds, maps)}"
end

main

File: 2023/day05/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function parseInput(text) {
  const sections = text.trim().split('\n\n');

  // Parse seeds
  const seeds = sections[0].split(': ')[1].split(' ').map(Number);

  // Parse maps
  const maps = [];
  for (let i = 1; i < sections.length; i++) {
    const lines = sections[i].trim().split('\n');
    const ranges = [];
    for (let j = 1; j < lines.length; j++) {
      const [dstStart, srcStart, length] = lines[j].split(' ').map(Number);
      ranges.push({ dstStart, srcStart, length });
    }
    maps.push(ranges);
  }

  return { seeds, maps };
}

function applyMap(value, ranges) {
  for (const { dstStart, srcStart, length } of ranges) {
    if (value >= srcStart && value < srcStart + length) {
      return dstStart + (value - srcStart);
    }
  }
  return value;
}

function seedToLocation(seed, maps) {
  let value = seed;
  for (const mapRanges of maps) {
    value = applyMap(value, mapRanges);
  }
  return value;
}

function part1(seeds, maps) {
  return Math.min(...seeds.map(seed => seedToLocation(seed, maps)));
}

function applyMapToRanges(inputRanges, mapRanges) {
  const result = [];

  for (const [start, end] of inputRanges) {
    let remaining = [[start, end]];

    for (const { dstStart, srcStart, length } of mapRanges) {
      const srcEnd = srcStart + length;
      const newRemaining = [];

      for (const [rStart, rEnd] of remaining) {
        // Part before the map range (unmapped)
        if (rStart < srcStart) {
          newRemaining.push([rStart, Math.min(rEnd, srcStart)]);
        }

        // Part within the map range (mapped)
        const overlapStart = Math.max(rStart, srcStart);
        const overlapEnd = Math.min(rEnd, srcEnd);
        if (overlapStart < overlapEnd) {
          const offset = dstStart - srcStart;
          result.push([overlapStart + offset, overlapEnd + offset]);
        }

        // Part after the map range (unmapped)
        if (rEnd > srcEnd) {
          newRemaining.push([Math.max(rStart, srcEnd), rEnd]);
        }
      }

      remaining = newRemaining;
    }

    // Any remaining parts are unmapped (identity)
    result.push(...remaining);
  }

  return result;
}

function part2(seeds, maps) {
  // Convert seeds to ranges
  let ranges = [];
  for (let i = 0; i < seeds.length; i += 2) {
    const start = seeds[i];
    const length = seeds[i + 1];
    ranges.push([start, start + length]);
  }

  // Apply each map to the ranges
  for (const mapRanges of maps) {
    ranges = applyMapToRanges(ranges, mapRanges);
  }

  // Find minimum start of any range
  return Math.min(...ranges.map(([start]) => start));
}

const inputPath = join(__dirname, '..', 'input.txt');
const text = readFileSync(inputPath, 'utf-8');

const { seeds, maps } = parseInput(text);

console.log('Part 1:', part1(seeds, maps));
console.log('Part 2:', part2(seeds, maps));
